DietPi-NordVPN | Add an update function

Joulinar opened the pull request:
**Status**:  Ready
- [x] tested general update procedure on RPi 32bit
- [x] tested with NordVPN account details available 

**Reference**: https://dietpi.com/phpbb/viewtopic.php?f=11&t=8221

**Commit list/description**:
- DietPi-NordVPN | Add a function to update NordVPN server list

Joulinar committed:
DietPi-NordVPN | Add an update function 

DietPi-NordVPN | Add a function to update NodVPN server list

Joulinar committed:
Update CHANGELOG.txt

Joulinar committed:
typo

Joulinar committed:
fix more typo

MichaIng commented:
Awesome, I'll run some tests tomorrow but LGTM so far. I was thinking that "Applying" the settings would be required again since the active config file is overwritten as well. But actually that should be clear? Or, if we do it, only of user and password have been set before to not apply dummy entries.

MichaIng commented on dietpi/misc/dietpi-nordvpn:
```diff
@@ -242,6 +242,7 @@ _EOF_
 			(( $NORDVPN_CONNECTED )) && G_WHIP_MENU_ARRAY+=('Disconnect' '')
 
 		fi
+		G_WHIP_MENU_ARRAY+=('Update' ': This will refresh NordVPN server list')
```
Maybe: "This updates the NordVPN server list"?
It looks a bid funny when "Refresh" "Update"s something while "Update" "refresh"es something 😄.

Then maybe above as well:
```
'Refresh' ': Refresh VPN connection status'
```

Joulinar replied:
yeah I know. I was struggling with the wording quite a lot. Maybe we would need to swap. `Update` will update VPN status while `Refresh` will refresh server list

MichaIng replied:
Also "update" sounds more like a real change compared to "refresh"?

Joulinar replied:
Well you could name it `Status Update` and `Server Liste Refresh`

Looks like `chicken` or `egg` question 🤣

MichaIng commented on dietpi/misc/dietpi-nordvpn:
```diff
@@ -333,6 +334,13 @@ _EOF_
+
+				G_EXEC cd /tmp/$G_PROGRAM_NAME
+				G_EXEC curl -sSfL https://downloads.nordcdn.com/configs/archives/servers/ovpn.zip -o 171.zip
+				G_EXEC unzip -o 171.zip -d/etc/openvpn
```
Maybe remove `/etc/openvpn/ovpn_tcp` and `/etc/openvpn/ovpn_udp` to also remove possibly obsolete/removed servers from the list? Or too risky if admins place custom configs inside?

Joulinar replied:
what about removing all `*.nordvpn.com.udp.ovpn` files?

MichaIng replied:
Good idea:
```
G_EXEC_DESC='Removing old NordVPN configs' G_EXEC rm -Rf /etc/openvpn/ovpn_{tcp,udp}/*.nordvpn.com.{tcp,udp}.ovpn
```
The problem is that `*` will expand to every single file being added to the `rm` command, so that output is veeeery long 😄. Hence the G_EXEC_DESC to shorten the output.

Joulinar commented:
Unfortunately I'm missing a NordVPN account. Therefore I'm not able to test it fully 😢 
So indeed some testing required how it behave having NodVPN configured already.

Joulinar committed:
add removing NordVPN server config files

Joulinar committed:
change wording

MichaIng commented on dietpi/misc/dietpi-nordvpn:
```diff
@@ -336,6 +336,7 @@ _EOF_
 
 			elif [[ $G_WHIP_RETURNED_VALUE == 'Update' ]]; then
 
+				G_EXEC_DESC='Removing old NordVPN server configs' G_EXEC rm -Rf /etc/openvpn/ovpn_{tcp,udp}/*.nordvpn.com.{tcp,udp}.ovpn
```
Let's do this only after successful download of the new archive. E.g. if something has changed, the download URL or the specific CDN target server is down, user will not end up with broken NordVPN, but can continue with the previous list.

Joulinar replied:
Ah that make sense. But than we would need to analyze the result of `G_EXEC curl -sSfL https://downloads.nordcdn.com/configs/archives/servers/ovpn.zip -o 171.zip`. Isn't it? How would you do that? Checking if the zip file exist? Or is there a return code given that can be used?

MichaIng commented:
I'll do a check later. Just thinking: Automatically re-applying settings, if everything required has been set, would be possible, basically the following check can be used: https://github.com/MichaIng/DietPi/blob/2961716d1bf9989f3f50e533e65917dcf9d594da/dietpi/misc/dietpi-nordvpn#L119
And of course it should be assured that the previously selected server (.ovpn file) is still available. If any of both is not true, the NORDVPN_SERVER variable should be unset to clearly show that it needs to be re-selected via empty value in the menu entry.

A different approach would be to preserve the currently active server config, e.g. storing the file to backup location and restore after list update. However that would break the use case to update the list in case of obsolete/removed server, outdated key, generally broken config and such. Probably best to keep the list update being a complete refresh of the .ovpn files.
____________
Okay I think doing less is better than doing too much here:
- We add an additional `G_WHIP_MSG` warning that the list update resets the server choice and all .ovpn files and that the server hence must be re-selected from the list + "Apply"ed to have all setup again. @Joulinar what you think?

Joulinar commented:
Yes personally I would keep thinks smart and just update the server list + inform user to enter personal settings again. I mean this will not happen that often. And just to fill user + password + select a server is not a big deal.

MichaIng commented:
Username and password (and port) would not need to be re-set, even when exiting, as those are stored in a dedicated settings file 😃. Only the server would need to be re-selected which makes sense if we are not sure whether the same server is even available.

A propos same server and reasons to update the server list: Would it be awesome to have a menu entry that pings the server (at best multiple times) and presents the average response time, to compare multiple servers? Or even select multiple servers, run some pings on all of them and nicely present the results (min, max, average) in a list? Okay that's a different topic, just wanted to have that written down somewhere 😄.

Joulinar commented:
> A propos same server and reasons to update the server list: Would it be awesome to have a menu entry that pings the server (at best multiple times) and presents the average response time, to compare multiple servers? Or even select multiple servers, run some pings on all of them and nicely present the results (min, max, average) in a list? Okay that's a different topic, just wanted to have that written down somewhere 😄.

oh yeah something for another project 😄 

> Username and password (and port) would not need to be re-set, even when exiting, as those are stored in a dedicated settings file 😃. Only the server would need to be re-selected which makes sense if we are not sure whether the same server is even available.

Even more simple. Just give the user a popup information about cleaning and re-downloading all server. This as well will clear the selected server and a new server needs to be chosen.

Joulinar committed:
Update dietpi-nordvpn

change order of activities, download the zip archive first, before cleaning old files

Commit:
TEST

Commit:
Revert "TEST"

This reverts commit f7ba824442dd880a8e70c40bac6ffb06ee33e26f.

Joulinar commented:
Sorry for the test commit. Was playing around with GitHub Desktop

btw, I changed order of activities and included the message box. However I noticed that the selected server is not cleared during refresh. Should we clear the selected server manually?

Joulinar commented:
@MichaIng 
are you able to test the full stack? I mean with having user+pw entered and a server selected. To see if it still working after refresh of server list.

MichaIng committed:
Merge branch 'dev' into NordVPN

MichaIng committed:
v6.34

+ DietPi-NordVPN | Unset the server choice when updating the server list, so hitting "Apply" prompts an info in the first place if the previously selected server is not available anymore.
+ DietPi-NordVPN | Apply new coding standard

MichaIng committed:
v6.34

+ DietPi-Patch | Syntax

Joulinar commented:
uh you update whole coding :)

MichaIng committed:
v6.34

+ DietPi-NordVPN | Syntax

MichaIng commented:
Yes, I'm going to remove the empty lines around if-then-else, for/while/until loops and such, they are a bid nasty since good editors start with a line indentation that matches the prior indentation, so those break the fluent code writing.

At least I want to check the effect of this on "beautify"/overview of the code, see if things can still be found as quickly as before, or if there are negatives.
____________
Okay, the script works fine. I turned the warning about the need to reselect the server into a yes/no dialog. Especially when users edited the ovpn file(s) manually, the should be able to cancel the step to e.g. backup their customisation first. Also the server variable is unset now, so hitting Apply without choosing again the server cannot lead to the attempt of applying a non-existent server. There was already a dialog to cancel "Apply" if not all variables are set (user, password, server).
